src/lib/common/InteractiveLayer.svelte

```diff
@@ -0,0 +1,116 @@
+<script lang="ts">
+  import {
+    Popup,
+    type MapGeoJSONFeature,
+    type MapLayerMouseEvent,
+  } from "maplibre-gl";
+  import { createEventDispatcher, onDestroy, onMount } from "svelte";
+  import { map } from "../../stores";
+
+  const dispatch = createEventDispatcher<{
+    click: MapGeoJSONFeature;
+  }>();
+
+  // This component manages hovering and clicking on objects in one layer. The
+  // caller must create and manage the source and layer. The source features must
+  // have numeric IDs defined. The caller can change style based on
+  // ["feature-state", "hover"] using the hoveredToggle helper.
+  export let layer: string;
+
+  // Optionally, display a popup for the hovered feature. The callback receives
+  // the hovered feature (noting that array and object properties are
+  // stringified). The tooltip contents will be wrapped in a govuk-prose div.
+  export let tooltip: ((feature: MapGeoJSONFeature) => string) | null = null;
+
+  // Toggle visibility by changing this
+  export let show = true;
+
+  // The currently hovered feature is exposed here. Note that any array or
+  // object properties get stringified here; this is not equivalent to the raw
+  // feature passed into the source.
+  export let hoveredFeature: MapGeoJSONFeature | null = null;
+
+  let popup = new Popup({
+    closeButton: false,
+    closeOnClick: false,
+    maxWidth: "none",
+  });
+
+  onMount(() => {
+    $map.on("mousemove", layer, onMouseMove);
+    $map.on("mouseleave", layer, onMouseLeave);
+    $map.on("click", layer, onClick);
+  });
+
+  onDestroy(() => {
+    $map.off("mousemove", layer, onMouseMove);
+    $map.off("mouseleave", layer, onMouseLeave);
+    $map.off("click", layer, onClick);
+    unhover();
+    popup.remove();
+  });
+
+  // Show/hide the layer
+  $: {
+    if ($map.getLayer(layer)) {
+      $map.setLayoutProperty(layer, "visibility", show ? "visible" : "none");
+    }
+  }
+
+  function onMouseMove(e: MapLayerMouseEvent) {
+    let features = e.features ?? [];
+    if (features.length == 0) {
+      return;
+    }
+    if (!features[0].id) {
+      console.log(
+        `Features in the ${layer} layer don't have IDs; InteractiveLayer won't work`
+      );
+    }
+
+    // This event will happen even if the feature is beneath something in a
```

Comment:
Asked for other ideas at https://osmus.slack.com/archives/C01G3D28DAB/p1690468687703299